csharp/src/SeedLang.Common/Range.cs

```diff
@@ -12,10 +12,34 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
+using System;
+
 namespace SeedLang.Common {
-  // Represents a range in the source code. Here, the source code can be either a plain text source
-  // code, or a SeedBlock module.
-  public class Range {
-    // Todo: implement the Range class.
+  // The base class of all the concrete code range classes.
+  public abstract class Range : IEquatable<Range> {
+    // Returns if the range is empty. A range can be empty when a diagnostic cannot be associated to
+    // a particular code position.
+    public abstract bool IsEmpty();
+
+    public abstract override int GetHashCode();
+
+    public abstract override string ToString();
```

Comment:
The GetHashCode and ToString are already defined in Object. I guess The intention to redefine it here is to hide the default implement and make sure all subclasses implement it. Then it's fine.

Reply:
GetHashCode() is for IEquality<>. The reason that I implement IEquality in the base class (the equality implementation of the Range/TextRange/BlockRange classes are slightly different from the example given by [How to define value equality for a type](https://docs.microsoft.com/en-us/dotnet/csharp/programming-guide/statements-expressions-operators/how-to-define-value-equality-for-a-type#class-example). The subclass like TextRange and BlockRange still accept a base class type as its argument of Equals. This provides the flexibility that we can compare the instances of the two different sub-types with "==" or "!=".

Reply:
ToString() is kept to hide the default implement as enforce all subclasses to provide a customized implementation.